lib/galaxy/workflow/modules.py

```diff
@@ -104,6 +118,126 @@ def __str__(self):
+    when_expression = step.when_expression
+    value_from_expressions = {}
+
+    if execution_state:
+        for key in execution_state.inputs.keys():
+            step_input = step.inputs_by_name.get(key)
+            if step_input and step_input.value_from is not None:
+                value_from_expressions[key] = step_input.value_from
+
+    if not value_from_expressions and when_expression is None:
+        return {}
+
+    hda_references = []
+
+    def to_cwl(value):
+        element_identifier = None
+        if isinstance(value, model.DatasetCollectionElement) and value.hda:
+            element_identifier = value.element_identifier
+            value = value.hda
+        if isinstance(value, model.HistoryDatasetAssociation):
+            # I think the following two checks are needed but they may
+            # not be needed.
+            if not value.dataset.in_ready_state():
+                why = "dataset [%s] is needed for valueFrom expression and is non-ready" % value.id
+                raise DelayedWorkflowEvaluation(why=why)
+            if not value.is_ok:
+                raise FailWorkflowEvaluation(
+                    why=InvocationFailureDatasetFailed(
+                        reason=FailureReason.dataset_failed, hda_id=value.id, workflow_step_id=step.id
+                    )
+                )
+            if value.ext == "expression.json":
+                with open(value.file_name) as f:
+                    # OUR safe_loads won't work, will not load numbers, etc...
+                    return json.load(f)
+            else:
+                hda_references.append(value)
+                properties = {
+                    "class": "File",
+                    "location": "step_input://%d" % len(hda_references),
+                }
+                set_basename_and_derived_properties(
+                    properties, value.dataset.created_from_basename or element_identifier or value.name
+                )
+                return properties
+        elif hasattr(value, "collection"):
+            collection = value.collection
+            if collection.collection_type == "list":
+                return [to_cwl(dce) for dce in collection.dataset_elements]
+            else:
+                # Could be record or nested lists
+                rval = {}
+                for element in collection.elements:
+                    rval[element.element_identifier] = to_cwl(element.element_object)
+                return rval
+        else:
+            return value
+
+    def from_cwl(value):
+        # TODO: turn actual files into HDAs here ... somehow I suppose. Things with
+        # file:// locations for instance.
+        if isinstance(value, dict) and "class" in value and "location" in value:
+            if value["class"] == "File":
+                # This is going to re-file -> HDA this each iteration I think, not a good
+                # implementation.
+                return progress.raw_to_galaxy(value)
```

Comment:
This is nothing right? How did we let this in 😆. It seems it hasn't caused any harm yet!